protocols/cosi/cosi.go

```diff
@@ -0,0 +1,373 @@
+	*sda.Node
+	// TreeNodeId cached
+	treeNodeId uuid.UUID
+	// the cosi struct we use (since it is a cosi protocol)
+	// Public because we will need it from other protocols.
+	Cosi *cosi.Cosi
+	// the message we want to sign typically given by the Root
+	message []byte
+	// The channel waiting for Announcement message
+	announce chan chanAnnouncement
+	// the channel waiting for Commitment message
+	commit chan chanCommitment
+	// the channel waiting for Challenge message
+	challenge chan chanChallenge
+	// the channel waiting for Response message
+	response chan chanResponse
+	// the channel that indicates if we are finished or not
+	done chan bool
+	// temporary buffer of commitment messages
+	tempCommitment []*CosiCommitment
+	// lock associated
+	tempCommitLock *sync.Mutex
+	// temporary buffer of Response messages
+	tempResponse []*CosiResponse
+	// lock associated
+	tempResponseLock *sync.Mutex
+	// hooks related to the various phase of the protocol.
+	// XXX NOT DEPLOYED YET / NOT IN USE.
+	// announcement hook
+	announcementHook AnnouncementHook
+	commitmentHook   CommitmentHook
+	challengeHook    ChallengeHook
+	DoneCallback     func(chal abstract.Secret, response abstract.Secret)
+}
+
+// NewProtocolCosi returns a ProtocolCosi with the round set and the
+// node set with the right channels.
+// If the round = nil, then we will create a default RoundCosi type instead.
+// Use this function like this:
+// ```
+// round := NewRound****()
+// fn := func(n *sda.Node) sda.ProtocolInstance {
+//      pc := NewProtocolCosi(round,n)
+//		return pc
+// }
+// sda.RegisterNewProtocolName("cothority",fn)
+// ```
+func NewProtocolCosi(node *sda.Node) (*ProtocolCosi, error) {
+	var err error
+	pc := &ProtocolCosi{
+		Cosi:             cosi.NewCosi(node.Suite(), node.Private()),
+		Node:             node,
+		done:             make(chan bool),
+		tempCommitLock:   new(sync.Mutex),
+		tempResponseLock: new(sync.Mutex),
+	}
+	// Register the three channels we want to register and listens on
+	// By passing pointer = automatic instantiation
+	node.RegisterChannel(&pc.announce)
+	node.RegisterChannel(&pc.commit)
+	node.RegisterChannel(&pc.challenge)
+	node.RegisterChannel(&pc.response)
+
+	// start the routine that listens on these channels
+	go pc.listen()
+	return pc, err
+}
+
+// NewRootProtocolCosi is used by the root to collectively sign this message
+// (vanilla version of the protocol where no contributions are done)
+func NewRootProtocolCosi(msg []byte, node *sda.Node) (*ProtocolCosi, error) {
+	pc, err := NewProtocolCosi(node)
+	pc.message = msg
+	return pc, err
+}
+
+// Start() will call the announcement function of its inner Round structure. It
+// will pass nil as *in* message.
+func (pc *ProtocolCosi) Start() error {
+	return pc.StartAnnouncement()
+}
+
+// Dispatch is not used, and already panics because it's DEPRECATED.
+func (pc *ProtocolCosi) Dispatch(msgs []*sda.SDAData) error {
```

Comment:
We should really take these away or keep them for the channel-listening-goroutine. At your convenience.